### DISCUSSION TITLE
How to reliably await VM boot?

### DISCUSSION THREAD

**Discussion by kanpov** [firecracker-microvm/firecracker#4708] [General]
For now, I haven't found a way to do this, as neither using a config file nor sending InstanceStart block until the VM has started. So it's either a hardcoded and hardware-specific delay, or analyzing the TTY which is a hack and doesn't work with daemonization where pipes are redir-d to /dev/null.

Is there a way to do this better?

**Comment by xmarcalx** [comment]
Hi @kanpov, 

Thanks for the message. 
There is no "standard" way to do it. The basic architecture model is to have an agent within the microVM which is able to audit the status of the VM (also the meaning on VM boot can be subjective application from application) and then try to communicate with the external world using for example using vsock, or networking.

From a naive point of view and just as example: if you run an SSH server within the microVM, you may consider that a VM booted if you can SSH into it. 

Let me know if it helps.

Cheers,
Marco

**Reply by kanpov** [reply]
> Hi @kanpov,
> 
> Thanks for the message. There is no "standard" way to do it. The basic architecture model is to have an agent within the microVM which is able to audit the status of the VM (also the meaning on VM boot can be subjective application from application) and then try to communicate with the external world using for example using vsock, or networking.
> 
> From a naive point of view and just as example: if you run an SSH server within the microVM, you may consider that a VM booted if you can SSH into it.
> 
> Let me know if it helps.
> 
> Cheers, Marco

How exactly would this agent (presumably a vsock binder) be started when the VM is booted? Wouldn't this involve, once again, having to await boot? The SSH route is not acceptable in many of my cases since I often use VMs without internet access.

**Comment by kanpov** [comment]
I figured out how to do this:
inject the GA (guest agent) as a precompiled binary into the rootfs, configure the init system (systemd) to invoke the binary on boot, await a host-initiated connection to the GA.
